lib/gssapi/krb5/store_cred.c

```diff
@@ -89,38 +91,41 @@ _gsskrb5_store_cred(OM_uint32         *minor_status,
 	return GSS_S_FAILURE;
     }
 
-    ret = krb5_cc_default(context, &def_ccache);
-    if (ret == 0) {
-        def_type = krb5_cc_get_type(context, def_ccache);
-        krb5_cc_close(context, def_ccache);
+    major_status = __gsskrb5_cred_store_find(minor_status, cred_store,
+					     "ccache", &cs_ccache_name, 0);
+    if (GSS_ERROR(major_status)) {
+	HEIMDAL_MUTEX_unlock(&cred->cred_id_mutex);
+	return major_status;
     }
-    def_ccache = NULL;
 
-    /* write out cred to credential cache */
-    ret = krb5_cc_cache_match(context, cred->principal, &id);
-    if (ret) {
-        if (default_cred) {
-            ret = krb5_cc_default(context, &id);
-            if (ret) {
-                HEIMDAL_MUTEX_unlock(&cred->cred_id_mutex);
-                *minor_status = ret;
-                return GSS_S_FAILURE;
-            }
-        } else {
-            if (def_type == NULL ||
-                !krb5_cc_support_switch(context, def_type)) {
-                HEIMDAL_MUTEX_unlock(&cred->cred_id_mutex);
-                *minor_status = 0;      /* XXX */
-                return GSS_S_NO_CRED;   /* XXX */
-            }
-            ret = krb5_cc_new_unique(context, def_type, NULL, &id);
-            if (ret) {
-                HEIMDAL_MUTEX_unlock(&cred->cred_id_mutex);
-                *minor_status = ret;
-                return GSS_S_FAILURE;
-            }
-            overwrite_cred = 1;
-        }
+    if (cs_ccache_name)
+	ret = krb5_cc_resolve(context, cs_ccache_name, &id);
+    else {
```

Comment:
This feels like a bug.  What if the app provided a non-empty cred store but didn't specify a ccache name?  We shouldn't then fallback on using the default ccache!  I think providing a non-empty cred store that doesn't specify a ccache name might be a bug.  I suppose we should check what MIT does and copy that.

Also, is there a difference between specifying a ccache key with a `NULL` name value and not specifying the key at all?

Reply:
New error `GSS_KRB5_S_G_UNKNOWN_CRED_STORE_ELEMENT`.

Re: specifying a key with a NULL value and not specifying it at all, good question. I'm going to say no, as there is not in MIT:

```c
OM_uint32
kg_value_from_cred_store(gss_const_key_value_set_t cred_store,
                         const char *type, const char **value)
{
    OM_uint32 i;

    if (value == NULL)
        return GSS_S_CALL_INACCESSIBLE_WRITE;
    *value = NULL;

    if (cred_store == GSS_C_NO_CRED_STORE)
        return GSS_S_COMPLETE;

    for (i = 0; i < cred_store->count; i++) {
        if (strcmp(cred_store->elements[i].key, type) == 0) {
            if (*value != NULL)
                return GSS_S_DUPLICATE_ELEMENT;
            *value = cred_store->elements[i].value;
        }
    }

    return GSS_S_COMPLETE;
}
```